thetis/solver.py

```diff
@@ -870,37 +865,30 @@ def create_equations(self):
                                                      self.fields.baroc_head_3d,
                                                      bottom_to_top=False,
                                                      average=False,
-                                                     bathymetry=self.fields.bathymetry_3d,
-                                                     elevation=self.fields.elev_cg_3d)
+                                                     bathymetry=self.fields.bathymetry_2d.view_3d,
+                                                     elevation=self.fields.elev_cg_2d.view_3d)
             self.int_pg_calculator = momentum_eq.InternalPressureGradientCalculator(
-                self.fields, self.options,
+                self.fields, self.fields.bathymetry_2d.view_3d, self.options,
                 self.bnd_functions['momentum'],
                 solver_parameters=self.options.timestepper_options.solver_parameters_momentum_explicit)
         self.extract_surf_dav_uv = SubFunctionExtractor(self.fields.uv_dav_3d,
                                                         self.fields.uv_dav_2d,
                                                         boundary='top', elem_facet='top',
                                                         elem_height=self.fields.v_elem_size_2d)
-        self.copy_elev_to_3d = ExpandFunctionTo3d(self.fields.elev_2d, self.fields.elev_3d)
-        self.copy_elev_cg_to_3d = ExpandFunctionTo3d(self.fields.elev_cg_2d, self.fields.elev_cg_3d)
         self.copy_uv_dav_to_uv_dav_3d = ExpandFunctionTo3d(self.fields.uv_dav_2d, self.fields.uv_dav_3d,
                                                            elem_height=self.fields.v_elem_size_3d)
         self.copy_uv_to_uv_dav_3d = ExpandFunctionTo3d(self.fields.uv_2d, self.fields.uv_dav_3d,
                                                        elem_height=self.fields.v_elem_size_3d)
-        self.uv_mag_solver = VelocityMagnitudeSolver(self.fields.uv_mag_3d, u=self.fields.uv_3d)
         self.mesh_updater = ALEMeshUpdater(self)
 
         if self.options.use_smagorinsky_viscosity:
-            self.smagorinsky_diff_solver = SmagorinskyViscosity(self.fields.uv_p1_3d, self.fields.smag_visc_3d,
+            self.smagorinsky_diff_solver = SmagorinskyViscosity(self.fields.uv_3d, self.fields.smag_visc_3d,
```

Comment:
Here, you also seem to drop the use of a P1CG-projected velocity. If I look at the `SmagorinskyViscosity` code I can believe that works with `weak_form=True` - but the logic here seems to only use the weak form for P0. For P0 it's obvious that the non-"weak form" case doesn't work, as the derivatives are zero - but for higher order DG (P1DG), the solution may still be piecewise constant, e.g. in a backwards facing step and I presume you still require viscosity at the shear interface?

Reply:
Yeah, that's right. In the 3D model we're always using a discontinuous velocity field so it's indeed better to always use the weak form. I've changed this now.